### ISSUE TITLE
Defining Schemas in Haskell?

### ISSUE THREAD

**Issue by pheaver** [seagreen/hjsonschema#43]
I'm interested in defining my schemas in Haskell and hjsonschema, and publishing by rendering out with ToJSON. I'd rather maintain my schema in a type-safe language than a huge amount of JSON.

As I understand it, this is certainly possible, but there's not much in hjsonschema to make it easier. I'm be interested in defining a set of DSL utilities for this.

I'm interested in what you think of this use case, whether it would be worth supporting in hjsonschema, and any other thoughts you may have about implementing it.

**Comment by seagreen** [comment]
Hey @pheaver, thanks for the question.

First let's cover what we have, which isn't much. [JSONSchema.Draft4.Schema](https://github.com/seagreen/hjsonschema/blob/master/src/JSONSchema/Draft4/Schema.hs) provides a data type that's isomorphic to valid JSON Schemas. This let's you write JSON Schemas using Haskell records instead of JSON:
```
$ stack ghci
> import qualified JSONSchema.Draft4 as D4
> import qualified JSONSchema.Validator.Draft4 as VAL
> let schema = D4.emptySchema { D4._schemaMaxItems = Just 2, D4._schemaType = Just (VAL.TypeValidatorString "array") }
> import qualified Data.ByteString.Lazy.Char8 as BC
> BC.putStrLn (encode schema)                                                                                                 
{"maxItems":2,"type":"array"}
```

What would you be interested in beyond this?

One thing I can think of it a schema record that can express any kind of validation that JSON Schemas can, but isn't exactly isomorphic to JSON Schema JSON. For instance, if you only care about validation than you don't care about the difference between `{"type": "object"}` and `{"type": ["object"]}`. Only allowing "type" to have an Array as it's value instead of a plain String would make these records simpler and easier to write than the one in `JSONSchema.Draft4.Schema`.

Another thing would be Haskell tools for building schemas using common patterns. For instance, it's often the case that you might want all the keys listed in `"properties"` to also go in `"required"`. While I think it would make sense for these tools to generate `JSONSchema.Draft4.Schema` schemas (and would be happy to help to write them!), I'm not sure their home should be in `hjsonschema` since this project wouldn't have a clearly limited scope.

**Comment by pheaver** [comment]
Yeah, I think this layer could benefit from being an abstraction around `Schema` and not isomorphic to it, but I understand something like that may not make sense in `hjsonschema`.

My question was inspired in part by [HaTeX](https://hackage.haskell.org/package/HaTeX), which defines utility functions for building up the AST (the `LaTeX`type, analogous to `Schema` in hjsonschema).

I wrote some pieces of a very basic DSL that demonstrates a little of what I had in mind, and an example that uses it to create a user schema. I think it demonstrates the bare minimum that one might want if they were to maintain a large schema defined in Haskell like this. It's just a thin abstraction around `Schema -> Schema`, and utility functions that define the basic `Schema -> Schema` functions. The `SchemaC` class is just a silly convenience thing for letting you either compose schema computations or immediately render them to a `Schema`, and isn't necessary for the rest of the utilities to work (I just had fun writing it).

```haskell
import qualified Data.HashMap.Lazy as H
import qualified Data.List.NonEmpty as NE
import           Data.Semigroup
import           Data.Set (Set)
import qualified Data.Set as Set
import           Data.Text (Text)
import           JSONSchema.Draft4
import           JSONSchema.Validator.Draft4
import           Prelude hiding (null)

uuid :: SchemaC s => s
uuid = build [ schemaType string
             , schemaPattern "^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$"
             ]

user :: SchemaC s => s
user = build [ draft4
             , schemaType object
             , definitions [ ("uuid", uuid) ]
             , properties [ ("userId", localRef "uuid")
                          , ("username", schemaType string)
                          , ("displayName", schemaType (string <> null))
                          ]
             , required ["userId", "username"]
             ]

instance Semigroup TypeValidator where
  (<>) x y = TypeValidatorArray (setFromTypeValidator x `Set.union` setFromTypeValidator y)

  sconcat xs = TypeValidatorArray $ Set.unions $ map setFromTypeValidator $ NE.toList xs

  stimes = stimesIdempotent

newtype SchemaF = SchemaF (Schema -> Schema)

class SchemaC s where
  liftS :: (Schema -> Schema) -> s

instance SchemaC Schema where
  liftS f = f emptySchema

instance SchemaC SchemaF where
  liftS = SchemaF

build :: SchemaC s => [SchemaF] -> s
build xs = liftS $ \s -> foldl (\s (SchemaF f) -> f s) s xs

schemaRef :: SchemaC s => Text -> s
schemaRef t = liftS $ \s -> s { _schemaRef = Just t }

localRef :: SchemaC s => Text -> s
localRef n = schemaRef ("#/definitions/" <> n)

schemaVersion :: SchemaC s => Text -> s
schemaVersion t = liftS $ \s -> s { _schemaVersion = Just t }

draft4 :: SchemaC s => s
draft4 = schemaVersion "http://json-schema.org/draft-04/schema#"

schemaType :: SchemaC s => TypeValidator -> s
schemaType tv = liftS $ \s -> s { _schemaType = Just tv }

schemaPattern :: SchemaC s => Text -> s
schemaPattern p = liftS $ \s -> s { _schemaPattern = Just p }

required :: SchemaC s => [Text] -> s
required ts = liftS $ \s -> s { _schemaRequired = Just (Set.fromList ts) }

properties :: SchemaC s => [(Text, Schema)] -> s
properties props = liftS $ \s -> s { _schemaProperties = Just (H.fromList props) }

definitions :: SchemaC s => [(Text, Schema)] -> s
definitions defs = liftS $ \s -> s { _schemaDefinitions = Just (H.fromList defs) }

object :: TypeValidator
object = TypeValidatorString "object"

string :: TypeValidator
string = TypeValidatorString "string"

null :: TypeValidator
null = TypeValidatorString "null"

```

**Comment by seagreen** [comment]
Hey @pheaver, sorry to get back to you so late.

Firstly, I really like the idea of making `TypeValidator` a Semigroup. That could be really useful in other contexts as well -- one of my goals to is be able to generate arbitrary JSON that passes validation for any particular schema, and I think I'll need to be able to concat validators together to do that.

Also, I like the idea of having `object`, `string`, etc. values for types. Actually, I should go farther and have `TypeValidator` not be based on `Text` at all -- I should have thought of this before.

Now, as to the DSL, I definitely see it as a syntactical improvement over writing schemas via records. Do you think it offers improvements in composability as well, or is it just meant to (significantly) ease the work of typing schemas out?

**Comment by pheaver** [comment]
Hi, no problem!

I've been adding onto that DSL and using it to define a big json schema, and I've been able to use it to compose schemas (afterall, the `build` just composes all the functions in the list).

That's cool that the `Semigroup` and other `TypeValidator` utilities might make it into hjsonschema! I'd agree that the rest is still probably too specific to my use case and outside the scope.

I realized that any function that updates a `Schema` should only ever add constraints (unlike the ones in the example I provided), and I've been changing them to behave like that while working on this. I've also been working on some support for defining named schemas of type `(Text, Schema)`, that render as a `$ref` to `"#/definitions/..."` and are included in the definitions list. Anyway, if I come up with anything else that I think could be useful, I'll bring it up :).
